orc/orcarm.c

```diff
@@ -1648,3 +1731,81 @@ orc_arm64_emit_mem (OrcCompiler *p, OrcArm64RegBits bits, OrcArm64Mem opcode,
 
   orc_arm_emit (p, code);
 }
+
+/** memory access instructions (pair)
+ *
+ *    3                   2                   1
+ *  1 0 9 8 7 6 5 4 3 2 1 0 9 8 7 6 5 4 3 2 1 0 9 8 7 6 5 4 3 2 1 0
+ * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ * |b 0|1 0 1|0| opt |L|     imm7    |   Rt2   |    Rn   |    Rt   |
+ * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ */
+
+#define arm64_code_mem_pair(b,opt,L,imm,Rt2,Rn,Rt) (0x40000000 | ((((b)==64)&0x1)<<31) | \
+    (((opt)&0x7)<<23) | (((L)&0x1)<<22) | (((imm)&0x7f)<<15) | (((Rt2)&0x1f)<<10) | \
+    (((Rn)&0x1f)<<5) | ((Rt)&0x1f))
+
+void
+orc_arm64_emit_mem_pair (OrcCompiler *p, OrcArm64RegBits bits, OrcArm64Mem opcode,
+    int opt, int Rt, int Rt2, int Rn, orc_uint64 val)
+{
+  orc_uint32 code;
+  orc_int32 imm;
+
+  static const char *insn_names[3] = {
+    "stp", "ldp"
+  };
+
+  char opt_rn[ARM64_MAX_OP_LEN];
+
+  if (opcode >= sizeof(insn_names)/sizeof(insn_names[0])) {
+    ORC_COMPILER_ERROR(p, "unsupported opcode %d", opcode);
+    return;
+  }
+
+  imm = (orc_int32) val;
+  memset (opt_rn, '\x00', ARM64_MAX_OP_LEN);
+
+  switch (opt) {
+    case 1: /** post-index */
+      snprintf (opt_rn, ARM64_MAX_OP_LEN, ", [%s], #%d", orc_arm64_reg_name(Rn, bits), imm);
+      break;
+    case 2: /** signed offset */
+      if (imm) {
+        snprintf (opt_rn, ARM64_MAX_OP_LEN, ", [%s, #%d]", orc_arm64_reg_name(Rn, bits), imm);
+      } else {
+        snprintf (opt_rn, ARM64_MAX_OP_LEN, ", [%s]", orc_arm64_reg_name(Rn, bits));
+      }
+      break;
+    case 3: /** pre-index */
+      snprintf (opt_rn, ARM64_MAX_OP_LEN, ", [%s, #%d]!", orc_arm64_reg_name(Rn, bits), imm);
+      break;
+    default:
+      ORC_COMPILER_ERROR(p, "unsupported variant %d\n", opt);
+      return;
+  }
+
+  if (bits == ORC_ARM64_REG_64) {
+    if (imm < -512*8 || imm > 504*8) {
```

Comment:
val was uint64. are you sure with this negative numbers?

Reply:
Oh.. val should be signed integer

Reply:
I changed the argument `orc_uint64 val` to `orc_in32 imm`, for simplicity.

Reply:
Is this imm == "imm7" (signed/7bit) in the figure? Then, now can it be -512*8 or 504*8?

Reply:
OMG. big mistake. I misread the manual.. :( Thanks!